core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala

```diff
@@ -0,0 +1,71 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+package kafka.controller
+
+import kafka.common.TopicAndPartition
+import kafka.server.ConfigType
+import org.apache.zookeeper.CreateMode
+import org.apache.zookeeper.KeeperException.Code
+
+object KafkaControllerZkUtils {
+  def getTopicPartitionStates(zookeeperClient: ZookeeperClient, contextPerPartition: Map[TopicAndPartition, Any]) = {
+    val getDataRequests = contextPerPartition.map { case (partition, ctx) =>
+      GetDataRequest(TopicPartitionStateZNode.path(partition), ctx)
+    }
+    retryUntilConnected(zookeeperClient, getDataRequests.toSeq).map(_.asInstanceOf[GetDataResponse])
+  }
+
+  def setTopicPartitionStates(zookeeperClient: ZookeeperClient,
+                              leaderIsrAndControllerEpochsAndContexts: Map[TopicAndPartition, (LeaderIsrAndControllerEpoch, Any)]) = {
+    val setDataRequests = leaderIsrAndControllerEpochsAndContexts.map { case (partition, (leaderIsrAndControllerEpoch, ctx)) =>
+      val path = TopicPartitionStateZNode.path(partition)
+      val data = TopicPartitionStateZNode.encode(leaderIsrAndControllerEpoch)
+      SetDataRequest(path, data, leaderIsrAndControllerEpoch.leaderAndIsr.zkVersion, ctx)
+    }
+    retryUntilConnected(zookeeperClient, setDataRequests.toSeq).map(_.asInstanceOf[SetDataResponse])
+  }
+
+  def createTopicPartitionStates(zookeeperClient: ZookeeperClient,
+                                 leaderIsrAndControllerEpochsAndContexts: Map[TopicAndPartition, (LeaderIsrAndControllerEpoch, Any)]) = {
+    val createRequests = leaderIsrAndControllerEpochsAndContexts.map { case (partition, (leaderIsrAndControllerEpoch, ctx)) =>
+      val path = TopicPartitionStateZNode.path(partition)
+      val data = TopicPartitionStateZNode.encode(leaderIsrAndControllerEpoch)
+      CreateRequest(path, data, null, CreateMode.PERSISTENT, ctx)
+    }
+    retryUntilConnected(zookeeperClient, createRequests.toSeq).map(_.asInstanceOf[CreateResponse])
```

Comment:
If a create operation retries due to Code.CONNECTIONLOSS, it's possible for the retry to receive a NodeExist error since the previous create may have succeeded. Perhaps on NodeExist error during retry, we can read the value back and only return NodeExist error if the value is different from that to be created.

Reply:
This falls in the category of suggestions that differ in current behavior. Again I'm not sure if we want to overload this PR with more than just porting existing behavior.